lib/ansible_tower_client/base_models/project.rb

```diff
@@ -17,5 +17,13 @@ def update
 
       api.project_updates.find(update['project_update'])
     end
+
+    def last_update_stdout(format = 'txt')
+      return if !related.respond_to?(:last_update) or related['last_update'].nil?
+
+      last_update_stdout_uri = URI("#{related['last_update']}stdout")
+      last_update_stdout_uri.query = URI.encode_www_form(format: format)
```

Comment:
Sorry, I was looking at api.rb... It still feels like there should be a URI building method elsewhere.

Reply:
I got inspired by https://github.com/Glutexo/ansible_tower_client_ruby/blob/afd22258e79804591a630b2ac5a2cd8d23c16c5d/lib/ansible_tower_client/base_models/job.rb#L11, only trying to avoid adding the query as a manually entered string. But yeah, we’re getting a bit un-DRY here. I’ll try to think out some an extraction to some helper class/method.

Reply:
Actually, the stdout capture can be retrieved from the JSON resource too. Thus there is no need to request `/stdout?format=txt`. We can simply have

```ruby
    def last_update
      return unless related['last_update'].present?
      ProjectUpdate.new(api, api.get(related['last_update']))
    end
```

and avoid the full URI machinery.

Using the same reasoning we can get rid of `Job::stdout` as well as unnecessary. Since this is an API gem, I’d rather open an issue for it instead of blindly removing it.

Reply:
I'm 👍 on less URI building.  I'm pretty new to this gem.  Does your `last_update` method return the same thing as what you have now?  Finding the most recent `project_updates` reads better to me.

Reply:
The new `last_update` method returns a whole `ProjectUpdate` object, the original `last_update_stdout` returned only the stdout string. This requires an update on the MIQ provider side too, but that’s no problem.

I can avoid using `related['last_update']` and fetch the most recent update manually by `api.project_updates.all('project' => project.id, 'order_by' => '-id').first`.